app/src/main/java/edu/wisc/ece/pinpoint/utils/FirebaseDriver.java

```diff
@@ -350,6 +394,83 @@ public Task<Void> deletePin(String pid) {
+                .addOnSuccessListener(complete -> db.collection("users").document(
+                        auth.getUid()).update("numFollowing", FieldValue.increment(1)));
+
+
+        // Add self to target's followers
+        db.collection("social")
+                .document(uid).update("followers", FieldValue.arrayUnion(auth.getUid()))
+                .addOnSuccessListener(complete -> db.collection("users").document(uid)
+                        .update("numFollowers", FieldValue.increment(1)));
+
+        pushActivityItem(new ActivityItem(auth.getUid(), uid, ActivityItem.ActivityType.FOLLOW, timestamp));
+    }
+
+    public void unfollowUser(String uid) {
+        if (auth.getUid() == null) {
+            throw new IllegalStateException("User must be logged in to follow an account");
+        }
+        followingIds.remove(uid);
+        // Remove target from following list
+        db.collection("social")
+                .document(auth.getUid()).update("following", FieldValue.arrayRemove(uid))
+                .addOnSuccessListener(complete -> db.collection("users").document(
+                        auth.getUid()).update("numFollowing", FieldValue.increment(-1)));
+
+        // Remove self from target's followers
+        db.collection("social")
+                .document(uid).update("followers", FieldValue.arrayRemove(auth.getUid()))
+                .addOnSuccessListener(complete -> db.collection("users").document(uid)
+                        .update("numFollowers", FieldValue.increment(-1)));
+    }
+
+    public Task<ActivityList> fetchActivity(String uid) {
+        if (auth.getUid() == null) {
+            throw new IllegalStateException("User must be logged in to fetch activity");
+        }
+        return db.collection("activity").document(uid).get().continueWith(task -> {
+            ActivityList activity;
+            if (!task.getResult().exists())
+                activity = new ActivityList();
+            else activity = task.getResult().toObject(ActivityList.class);
+            activityMap.put(uid, activity);
+            return activity;
+        });
+    }
+
+    public ActivityList getCachedActivity(String uid) {
+        if(activityMap.containsKey(uid))
+            return activityMap.get(uid);
+        return null;
+    }
+
+    public void pushActivityItem(ActivityItem activityItem) {
+        if (auth.getUid() == null) {
+            throw new IllegalStateException("User must be logged in to push activity");
+        }
+        String authorId = activityItem.getAuthor();
+        activityMap.get(authorId).add(activityItem);
+        // Create map to store field name and activity item
+        HashMap<String, Object> map = new HashMap<>();
+        map.put("activity",FieldValue.arrayUnion(activityItem));
+        // Push activity item to firebase
+        db.collection("activity").document(authorId).set(map, SetOptions.merge());
+    }
```

Comment:
This could also be made to be part of the batched write for follow. I feel like we don't need a whole method to push activity updates since it will always happen in coordination with following someone or adding a comment, so we should just create & push the specific activity types from within each of those methods. Then we can also use the batched write that will already be available.